### ISSUE TITLE
qa: search qa tests failing in CI

### ISSUE THREAD

**Issue by davejrt** [sourcegraph/sourcegraph#15178]
search tests are failing in the regression pipline in CI. These tests need to be updated or removed if they're no longer valid. 

The errors with the exception of one are due to strings not matching, which on face value just appears that the expected string needs to be changed because the tests are old/no longer valid. 

The other error is a timeout. 

Full log here:

https://buildkite.com/sourcegraph/regression/builds/95#4844d831-db41-4b9d-964a-8b6052edfff0/40-6968

**Comment by davejrt** [comment]
Update:

There appears to have been a regression introduced between 3.20 and 3.21 here. The following three tests are failing when run 3.21 and latest single server images. I have also replicated this deploying with docker-compose. These are always run on a fresh deployment with no existing data. 

Below are the three errors:

```
Visited routes:
/sign-in
/search


  38 passing (7m)
  3 failing

  1) Search regression test suite
       Search over a dozen repositories
         Global text search (repohasfile:copying), expect many results.:
     TimeoutError: waiting for function failed: timeout 30000ms exceeded
      at new WaitTask (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:394:34)
      at DOMWorld.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:316:16)
      at Frame.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:869:32)
      at Page.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1221:33)
      at Context.<anonymous> (src/regression/search.test.ts:196:31)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  2) Search regression test suite
       Search over a dozen repositories
         Global text search, fork:only, few results:
     TimeoutError: waiting for function failed: timeout 30000ms exceeded
      at new WaitTask (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:394:34)
      at DOMWorld.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:316:16)
      at Frame.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:869:32)
      at Page.waitForFunction (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1221:33)
      at Context.<anonymous> (src/regression/search.test.ts:234:31)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)

  3) Search regression test suite
       Search over a dozen repositories
         Search suggestions:
     TimeoutError: waiting for selector ".monaco-query-input-container .suggest-widget.visible" failed: timeout 30000ms exceeded
      at new WaitTask (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:394:34)
      at DOMWorld._waitForSelectorOrXPath (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:326:26)
      at DOMWorld.waitForSelector (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:309:21)
      at Frame.waitForSelector (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:801:51)
      at Page.waitForSelector (/home/davet/sourcegraph/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1215:33)
      at Context.<anonymous> (src/regression/search.test.ts:418:31)
```

**Comment by rvantonder** [comment]
Is there a docker image for the qa testing? I'd like to try things with the image manually, since I wasn't able to reproduce "No results" when I built an image from main branch, things appear to still work.

The output above only say timeout, so there might be something else going on.  Would like to try things.

**Comment by davejrt** [comment]
@rvantonder the QA tests run on the current candidate image, but for the sake of manually testing I have been pulling down the latest insiders images.

**Comment by rvantonder** [comment]
@davejrt I can confirm `repohasfile` works on the current `insiders` server image for a small repo, and I don't think there's a regression there. This suggests an issue with the test/infra. I wonder if the test repositories have changed and/or the search doesn't return or find results, I know that we don't pin to specific repos here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/regression/search.test.ts#L66. I will try cloning that set of repos and seeing if I can reproduce the exact case.

**Comment by davejrt** [comment]
@rvantonder I've started with a fresh `insiders` image with all repos in sourcegraph as well as those [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/regression/search.test.ts#L66) and get zero results. 

Can you confirm which repos you got results from and also that the search is only `repohasfile:copyfile` and it's a `regex` search?

**Comment by rvantonder** [comment]
So when I added all the repos I get no results as well. Even when I restart the image and wipe the repos, I get no results (so I can't replicate the success case before). This is bizarre. I'm going to backpedal here and actually go understand what's happening in the code :/

**Comment by pecigonzalo** [comment]
It seems like some of the tests are still failing :( https://buildkite.com/sourcegraph/qa/builds/527#4b9e3fb0-6858-4b0e-8e7c-e41efd1c9299/33-7062

```

sourcegraph-qa-test:   1) Search regression test suite
--
  | sourcegraph-qa-test:        Search over a dozen repositories
  | sourcegraph-qa-test:          Global text search, fork:only, few results:
  | sourcegraph-qa-test:      TimeoutError: waiting for function failed: timeout 30000ms exceeded
  | sourcegraph-qa-test:       at new WaitTask (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:394:34)
  | sourcegraph-qa-test:       at DOMWorld.waitForFunction (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:316:16)
  | sourcegraph-qa-test:       at Frame.waitForFunction (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:869:32)
  | sourcegraph-qa-test:       at Page.waitForFunction (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1221:33)
  | sourcegraph-qa-test:       at Context.<anonymous> (src/regression/search.test.ts:234:31)
  | sourcegraph-qa-test:       at runMicrotasks (<anonymous>)
  | sourcegraph-qa-test:       at processTicksAndRejections (internal/process/task_queues.js:93:5)
  | sourcegraph-qa-test:
  | sourcegraph-qa-test:   2) Search regression test suite
  | sourcegraph-qa-test:        Search over a dozen repositories
  | sourcegraph-qa-test:          Search suggestions:
  | sourcegraph-qa-test:      TimeoutError: waiting for selector ".monaco-query-input-container .suggest-widget.visible" failed: timeout 30000ms exceeded
  | sourcegraph-qa-test:       at new WaitTask (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:394:34)
  | sourcegraph-qa-test:       at DOMWorld._waitForSelectorOrXPath (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:326:26)
  | sourcegraph-qa-test:       at DOMWorld.waitForSelector (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:309:21)
  | sourcegraph-qa-test:       at Frame.waitForSelector (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/FrameManager.js:801:51)
  | sourcegraph-qa-test:       at Page.waitForSelector (/sourcegraph/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1215:33)
  | sourcegraph-qa-test:       at Context.<anonymous> (src/regression/search.test.ts:418:31)


```

**Comment by lguychard** [comment]
After #15922, the last failing test was `Global text search, fork:only, few results`.

This test expects more than 5 filematches to be returned, but in 3.22 only 4 filematches are returned in the initial results.

I found out that:
- This was already the case in 3.21 (92 results over 4 filematches returned, instead of 588 results over 14 filematches).
- The behaviour that was introduced in https://github.com/sourcegraph/sourcegraph/commit/9e0d8f8999b88c17b23f7c71f884a46f2c2c1960.
- The behaviour is not a _bug_, as it only impacts the number of results initially returned, but with a higher count the exhaustive result set is still the same between 3.20 and 3.21.

So my take is that we can adjust the expected number of filematches in the test to be 4 instead of 5.

**Comment by pecigonzalo** [comment]
> So my take is that we can adjust the expected number of filematches in the test to be 4 instead of 5.

@lguychard is there a way to generalize that? like expect `> 2` to allow that to change without breaking the tests. My concern is that if the tests are too specific, they will break often.

**Comment by rvantonder** [comment]
> like expect > 2 to allow that to change without breaking the tests. 

This is exactly how our backend integration tests are set up, and we have coverage for a lot of these e2e tests including `fork` behavior, and others I see in there. I thought they were removed when we introduced our backend integration tests but I guess not. I will do a pass to delete the e2e tests we no longer need or convert the remainder that are relevant to backend only.

**Comment by pecigonzalo** [comment]
That would be great @rvantonder! I think the more we can trim this test, the better. I suspected many of these were duplicate in places that are better/faster.

**Comment by rvantonder** [comment]
I don't think there are any more search tests failing from incorrect/flaky behavior. I see a failure still for:

> sourcegraph-qa-test:   20 passing (8m)
--
  | sourcegraph-qa-test:   1 failing
  | sourcegraph-qa-test:   1) Search regression test suite
  | sourcegraph-qa-test:        Search over a dozen repositories
  | sourcegraph-qa-test:          Search filters:
  | sourcegraph-qa-test:      Error: RetryOperation timeout occurred
  | sourcegraph-qa-test:       at RetryOperation.retry (/sourcegraph/node_modules/retry/lib/retry_operation.js:49:26)
  | sourcegraph-qa-test:       at RetryOperation._fn (/sourcegraph/node_modules/p-retry/index.js:64:20)

But this "search over a dozen repositories" test is probably failing because of a test setup failure (see https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/regression/search.test.ts#L121-129). External service, user? I'm not sure, not familiar with this test.

**Comment by pecigonzalo** [comment]
@rvantonder Thanks! Why do you think its related to the test setup? Im also not familiar with the test, but it seems to be defined here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/regression/search.test.ts#L253-279 I suspect if it was the `before` part, it will fail all tests of that set.

Do you know what that is testing for? Is that covered by some other test or integration test?

**Comment by pecigonzalo** [comment]
Im waiting for the CI environment to verify, but that test and all Search test seem to pass now on the `3.22` branch.

**Comment by pecigonzalo** [comment]
Yep, all tests passing on CI. https://buildkite.com/sourcegraph/qa/builds/571#300b9ee4-ae81-4407-a945-b4bc1afb166e

**Comment by lguychard** [comment]
@pecigonzalo reopening as I need to revert one of the fixes, which introduces other regressions: https://github.com/sourcegraph/sourcegraph/pull/15974

**Comment by pecigonzalo** [comment]
Fixed on #15978
